File: src/main/java/ru/org/linux/topic/TagTopicListController.java
package ru.org.linux.topic;

import org.apache.commons.lang.WordUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.View;
import org.springframework.web.servlet.view.RedirectView;
import org.springframework.web.util.UriComponentsBuilder;
import org.springframework.web.util.UriTemplate;
import ru.org.linux.section.Section;
import ru.org.linux.section.SectionService;
import ru.org.linux.site.Template;
import ru.org.linux.tag.TagService;
import ru.org.linux.user.UserTagService;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.util.List;

@Controller
public class TagTopicListController {
  private static final UriTemplate TAG_URI_TEMPLATE = new UriTemplate("/tag/{tag}");
  private static final UriTemplate TAGS_URI_TEMPLATE = new UriTemplate("/tags/{tag}");
  @Autowired
  private UserTagService userTagService;

  @Autowired
  private SectionService sectionService;

  @Autowired
  private TagService tagService;

  @Autowired
  private TopicListService topicListService;

  @Autowired
  private TopicPrepareService prepareService;

  public static String tagListUrl(String tag) {
    return TAG_URI_TEMPLATE.expand(tag).toString();
  }

  public static String tagsUrl(char letter) {
    return TAGS_URI_TEMPLATE.expand(letter).toString();
  }

  private String getTitle(@Nonnull String tag, @Nullable Section section) {
    if (section==null) {
      return WordUtils.capitalize(tag);
    } else {
      return WordUtils.capitalize(tag) + " (" + section.getName() + ")";
    }
  }

  @RequestMapping(value = "/tag/{tag}", method = {RequestMethod.GET, RequestMethod.HEAD})
  public ModelAndView tagFeed(
    HttpServletRequest request,
    HttpServletResponse response,
    @PathVariable String tag,
    @RequestParam(value = "offset", defaultValue = "0") int offset,
    @RequestParam(value = "section", defaultValue = "0") int sectionId
  ) throws Exception {
    ModelAndView modelAndView = new ModelAndView("tag-topics");

    Section section;

    if (sectionId!=0) {
      section = sectionService.getSection(sectionId);
      modelAndView.addObject("section", section);
    } else {
      section = null;
    }

    tagService.checkTag(tag);

    TopicListController.setExpireHeaders(response, null, null);

    String title = getTitle(tag, section);

    modelAndView.addObject("navtitle", title);
    modelAndView.addObject("ptitle", title);

    offset = topicListService.fixOffset(offset);

    List<Topic> topics = topicListService.getTopicsFeed(
            section,
            null,
            tag,
            offset,
            null,
            null
    );

    Template tmpl = Template.getTemplate(request);

    List<PersonalizedPreparedTopic> preparedTopics = prepareService.prepareMessagesForUser(
            topics,
            request.isSecure(),
            tmpl.getCurrentUser(),
            tmpl.getProf(),
            false
    );

    modelAndView.addObject(
            "messages",
            preparedTopics
    );

    modelAndView.addObject("offsetNavigation", true);

    modelAndView.addObject("tag", tag);
    modelAndView.addObject("section", sectionId);
    modelAndView.addObject("offset", offset);
    modelAndView.addObject("sectionList", sectionService.getSectionList());

    if (tmpl.isSessionAuthorized()) {
      modelAndView.addObject(
              "isShowFavoriteTagButton",
              !userTagService.hasFavoriteTag(tmpl.getCurrentUser(), tag)
      );

      modelAndView.addObject(
              "isShowUnFavoriteTagButton",
              userTagService.hasFavoriteTag(tmpl.getCurrentUser(), tag)
      );

      if (!tmpl.isModeratorSession()) {
        modelAndView.addObject(
                "isShowIgnoreTagButton",
                !userTagService.hasIgnoreTag(tmpl.getCurrentUser(), tag)
        );
        modelAndView.addObject(
                "isShowUnIgnoreTagButton",
                userTagService.hasIgnoreTag(tmpl.getCurrentUser(), tag)
        );
      }
    }

    modelAndView.addObject("counter", tagService.getCounter(tag));

    modelAndView.addObject("url", tagListUrl(tag));
    modelAndView.addObject("favsCount", userTagService.countFavs(tagService.getTagId(tag)));

    if (offset<200 && preparedTopics.size()==20) {
      modelAndView.addObject("nextLink", buildTagUri(tag, sectionId, offset + 20));
    }

    if (offset>=20) {
      modelAndView.addObject("prevLink", buildTagUri(tag, sectionId, offset - 20));
    }

    return modelAndView;
  }

  private String buildTagUri(String tag, int section, int offset) {
    UriComponentsBuilder builder = UriComponentsBuilder.fromUri(TAG_URI_TEMPLATE.expand(tag));

    if (section!=0) {
      builder.queryParam("section", section);
    }

    if (offset!=0) {
      builder.queryParam("offset", offset);
    }

    return builder.build().toUriString();
  }

  @RequestMapping(value = "/view-news.jsp", method = {RequestMethod.GET, RequestMethod.HEAD}, params = {"tag"})
  public View tagFeedOld(
          @RequestParam String tag
  ) {
    return new RedirectView(TagTopicListController.tagListUrl(tag));
  }
}

File: src/main/scala/ru/org/linux/topic/TagTopicListController.scala
/*
 * Copyright 1998-2023 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.topic

import org.springframework.http.HttpStatus
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.{ExceptionHandler, PathVariable, RequestMapping, RequestMethod, RequestParam, ResponseStatus}
import org.springframework.web.servlet.view.RedirectView
import org.springframework.web.servlet.{ModelAndView, View}
import org.springframework.web.util.{UriComponentsBuilder, UriTemplate}
import ru.org.linux.auth.AuthUtil.AuthorizedOpt
import ru.org.linux.group.GroupListDao
import ru.org.linux.section.{Section, SectionNotFoundException, SectionService}
import ru.org.linux.site.Template
import ru.org.linux.tag.{TagName, TagNotFoundException, TagPageController, TagService}
import ru.org.linux.user.UserTagService

import scala.concurrent.ExecutionContext.Implicits.global
import scala.compat.java8.FutureConverters.*
import java.util.concurrent.CompletionStage
import scala.concurrent.Future
import scala.jdk.CollectionConverters.{ListHasAsScala, SeqHasAsJava}
import scala.jdk.OptionConverters.RichOption

@Controller
object TagTopicListController {
  private val TagUriTemplate = new UriTemplate("/tag/{tag}")
  private val TagUriSectionTemplate = new UriTemplate("/tag/{tag}?section={section}")
  private val TagsUriTemplate = new UriTemplate("/tags/{tag}")

  def tagListUrl(tag: String): String = TagUriTemplate.expand(tag).toString

  def tagListUrl(tag: String, section: Section): String =
    TagUriSectionTemplate.expand(tag, Integer.valueOf(section.getId)).toString

  def tagsUrl(letter: Char): String = TagsUriTemplate.expand(letter.toString).toString

  private def buildTagUri(tag: String, section: Int, offset: Int) = {
    val builder: UriComponentsBuilder = UriComponentsBuilder.fromUri(TagUriTemplate.expand(tag))

    if (section != 0) {
      builder.queryParam("section", Integer.valueOf(section))
    }

    if (offset != 0) {
      builder.queryParam("offset", Integer.valueOf(offset))
    }

    builder.build.toUriString
  }
}

@Controller
class TagTopicListController(userTagService: UserTagService, sectionService: SectionService, tagService: TagService,
                             topicListService: TopicListService, prepareService: TopicPrepareService,
                             topicTagDao: TopicTagDao, groupListDao: GroupListDao) {

  private def getTitle(tag: String, section: Section) = s"${tag.capitalize} (${section.getName})"

  @RequestMapping(
    value = Array("/tag/{tag}"),
    method = Array(RequestMethod.GET, RequestMethod.HEAD),
    params = Array("section"))
  def tagFeed(@PathVariable tag: String,
               @RequestParam(value = "offset", defaultValue = "0") rawOffset: Int,
               @RequestParam(value = "section", defaultValue = "0") sectionId: Int
  ): CompletionStage[ModelAndView] = AuthorizedOpt { currentUserOpt =>
    TagName.checkTag(tag)

    val deadline = TagPageController.Timeout.fromNow

    val section = sectionService.getSection(sectionId)

    val countF = tagService.countTagTopics(tag = tag, section = Some(section), deadline = deadline)

    (tagService.getTagInfo(tag, skipZero = true) match {
      case Some(tagInfo) =>
        val forumMode = section.getId == Section.SECTION_FORUM  && currentUserOpt.exists(_.user.isAdministrator)

        // todo old/new
        // todo tag-topics-forum
        val modelAndView = new ModelAndView(if (forumMode) "tracker-new" else "tag-topics")

        modelAndView.addObject("section", section)

        modelAndView.addObject("tagTitle", tag.capitalize)
        modelAndView.addObject("ptitle", getTitle(tag, section))

        val offset = TopicListService.fixOffset(rawOffset)

        modelAndView.addObject("offsetNavigation", true)
        modelAndView.addObject("tag", tag)
        modelAndView.addObject("section", sectionId)
        modelAndView.addObject("offset", offset)

        val sections = topicTagDao.getTagSections(tagInfo.id).map(sectionService.idToSection)

        modelAndView.addObject("sectionList", sections.asJava)

        currentUserOpt.foreach { currentUser =>
          modelAndView.addObject("showFavoriteTagButton", !userTagService.hasFavoriteTag(currentUser.user, tag))
          modelAndView.addObject("showUnFavoriteTagButton", userTagService.hasFavoriteTag(currentUser.user, tag))

          if (!currentUser.moderator) {
            modelAndView.addObject("showIgnoreTagButton", !userTagService.hasIgnoreTag(currentUser.user, tag))
            modelAndView.addObject("showUnIgnoreTagButton", userTagService.hasIgnoreTag(currentUser.user, tag))
          }
        }

        val prof = Template.getTemplate.getProf

        val (preparedTopics, pageSize) = if (forumMode) {
          (groupListDao.getSectionListTopics(section, currentUserOpt.map(_.user).toJava,
            prof.getTopics, offset, prof.getMessages, tagInfo.id), prof.getTopics)
        } else {
          val topics = topicListService.getTopicsFeed(section, None, Some(tag), offset, None,
            20, currentUserOpt.map(_.user), noTalks = false, tech = false)

          (prepareService.prepareTopicsForUser(topics, currentUserOpt.map(_.user), prof, loadUserpics = false), 20)
        }

        modelAndView.addObject("messages", preparedTopics)

        modelAndView.addObject("url", TagTopicListController.tagListUrl(tag))
        modelAndView.addObject("favsCount", userTagService.countFavs(tagInfo.id))
        modelAndView.addObject("ignoreCount", userTagService.countIgnore(tagInfo.id))

        if (offset < TopicListService.MaxOffset && preparedTopics.size == pageSize) {
          modelAndView.addObject("nextLink", TagTopicListController.buildTagUri(tag, sectionId, offset + pageSize))
        }

        if (offset > pageSize) {
          modelAndView.addObject("prevLink", TagTopicListController.buildTagUri(tag, sectionId, offset - pageSize))
        }

        if (preparedTopics.isEmpty) {
          Future.successful(new ModelAndView("errors/code404"))
        } else {
          countF.map { count =>
            modelAndView.addObject("counter", count.getOrElse(0))
            modelAndView
          }
        }
      case None =>
        tagService.getTagBySynonym(tag).map { mainName =>
          Future.successful(new ModelAndView(new RedirectView(TagTopicListController.buildTagUri(mainName.name, sectionId, 0), false, false)))
        }.getOrElse(throw new TagNotFoundException())
    }).toJava
  }

  @RequestMapping(
    value = Array("/view-news.jsp"),
    method = Array(RequestMethod.GET, RequestMethod.HEAD),
    params = Array("tag"))
  def tagFeedOld(@RequestParam tag: String): View = new RedirectView(TagTopicListController.tagListUrl(tag))

  @ExceptionHandler(Array(classOf[SectionNotFoundException]))
  @ResponseStatus(HttpStatus.NOT_FOUND)
  def handleNotFoundException = new ModelAndView("errors/code404")
}
